pkg/webhook/pod/pod.go

```diff
@@ -65,13 +68,34 @@ func (a *podAnnotator) Handle(ctx context.Context, req admission.Request) admiss
 		return admission.Errored(http.StatusInternalServerError, err)
 	}
 
-	marshaledPod, _ := json.Marshal(pod)
+	// admission.PatchResponse generates a Response containing patches.
+	return patchPodChanges(originalPod, pod)
```

Comment:
Since we know all the fields we intent to control, making diff of the whole objects seems redundant. Not only that, onl line 66 we change the pod in-place. AllocatePodMac is a black box and we just assume in patchPodChanges that those fields are the only ones that can change. This PR does the job, but it extends the existing code, even though part of it may not be needed anymore.

In order to make the code-base maintainable, could you consider changing AllocatePodMac to it does not mutate the pod? e.g. it could return a map of interfaces and their assigned MACs. We can generate the jsonPatch based on this information alone. That would limit the function interface to minimum, make it possible to test it and judge the behavior based on the function signature alone.

Reply:
Can you elaborate exactly what you suggest to change? Do you mean only on pod's handler or also vm handler? 

In pods handler we only change the annotation, but in virtualmachine handler we add the interface's macAddress and also add finalizer. to common denominator is of the last 2 is the vm spec itself. Alternatively I can return a struct of all the changes (interfaces & finalizers) but I don't think it would make it more maintainable.
That's why instead of returning only the changes, and constructing the patches from them, I preferred in `patchPodChanges` and `patchVMChanges` to return the copy of the changed vm/pod and then extract exactly the needed changes from it.

Reply:
I would not even return the changes, just which MAC should be allocated to each VM. This way you'd be able to test the reading of the VM, figuring out whether we handle the allocation correctly.

The next function will get this map and apply it on a specific VM - return patches. I'm not sure I follow why'd you return finalizers. If MACs are allocated, you need a finalizer. If not, you don't need any. This can be decided by the second function.

I will not fight for this. Although I believe that having smaller functions with limited signatures would make the maintainability and testability easier.

Reply:
I see what you mean, but changing the logic in the way you suggest needs more careful examination in order to unharm the current logic. Can we further discuss and do the change you suggest in another PR?

Reply:
Fair enough